### ISSUE TITLE
Learning speed not as expected

### ISSUE THREAD

**Issue by ArtemiyFirsov** [SeanNaren/deepspeech.pytorch#118]
Hey guys, trying to run training with with default params on 4 x GTX 1080 16gb, which gives 44 teraflops.
On full dataset (about 1000 hours), it takes about 20 hours to train. It means, that if I want to train on 100.000 hours dataset, as originally suggested in DeepSpeech2 article, it will take about 12 weeks, which is too long, comparing with 2 weeks to Baidu did. 
They had 8 Titans, but I looked at characteristics, mine GPUs should quite the same in terms if performance, meaning I should get a month of learning, but where do 3 months come from, I do not get an idea. Can you please advice smth on how to speed it up?

**Comment by ryanleary** [comment]
Baidu's multi-gpu implementation and memory usage is highly optimized. This allows them to better leverage data parallelism and increase batch size. There are limits to the performance we'll be able to achieve using PyTorch.

**Comment by SeanNaren** [comment]
what @ryanleary said covers it all, sadly using Torch we've run into memory optimization and performance issues that Baidu spent a lot of time sussing out. Also 100K hours?? They are able to train on larger and larger datasets, but honestly anything within the 3-11k hours of audio range can train a decent model, then bootstrap a good ken LM model!

**Comment by ryanleary** [comment]
A couple other things I'd like to add:
1) Do you actually have 100k hours (and if so, wow...) or are you planning on doing some kind of synthesis?
2) Time to convergence could probably be improved in this implementation by getting batch normalization working. This too, however, would require pytorch changes. Though this is probably a lighter lift compared to the other things already mentioned.

**Comment by ArtemiyFirsov** [comment]
@SeanNaren @ryanleary thank you for your answers, really appreciated! 
considering Baidu's memory optimizations, here is there note about memory:
_These optimizations include a fast implementation of the CTC loss function on the GPU, and a custom memory allocator. We also use carefully integrated compute nodes and a custom implementation of all-reduce to accelerate inter-GPU communication._
So, we have warp-ctc library, that OK. 
We do not have custom memory allocator. My guess is that your lib is written in PyTorch, we need to improve Python memory allocation, there are some materials on that - http://www.evanjones.ca/memoryallocator/
Also, GPU memory allocation - https://github.com/canonizer/halloc

About 100K h dataset - my idea is to gather 10K h of audio and inject noise in it (very different noise). However, maybe i can get ~ 10K h and train different models for different situations on 20Kh with noise - like, telephone call, skype call, meeting, etc.  
Of course, 3-11k might by enough, but it is certainly not enough to cope with noise. Correct me if I am wrong.

And, BTW, can you please suggest any dataset to train KenLM model?
